move-vm-support/src/base58_address.rs

```diff
@@ -0,0 +1,183 @@
+//! Base58 address format converter for Substrate and Move accounts.
+
+use anyhow::{bail, Result};
+use move_core_types::account_address::AccountAddress;
+
+use super::PUB_KEY_LEN;
+
+// Maximum supported Base58 address length in bytes
+const BASE58_LEN: usize = PUB_KEY_LEN;
+
+/// Convert Base58 address string to Move address structure.
+/// In case if such conversion is not possible, return error.
+/// ```
+/// use move_vm_support::base58_address::base58_to_move_address;
+/// let substrate_address = "BH9wjrYwKi76DbU6LC1X4VEeevUTkaKgxGnaFNFaqZA";
+/// let move_address = base58_to_move_address(substrate_address).unwrap();
+/// assert_eq!(
+///    "0x02A212DE6A9DFA3A69E22387ACFBAFBB1A9E591BD9D636E7895DCFC8DE05F331",
+///   format!("{:#X}", move_address)
+/// );
+/// ```
+pub fn base58_to_move_address(base58: &str) -> Result<AccountAddress> {
+    // Decoded format: <address>
+    //  Size in bytes:    32
+    let decoded_base58 = bs58::decode(base58).into_vec()?;
+
+    // Check if the length is valid and figure out the address type length.
+    if decoded_base58.len() != BASE58_LEN {
+        bail!(
+            "unsupported base58 address length (base58): {}",
+            decoded_base58.len()
+        );
+    }
+
+    AccountAddress::from_bytes(decoded_base58).map_err(anyhow::Error::msg)
+}
+
+/// Convert Base58 address to Move address string.
+pub fn base58_to_move_address_string(base58: &str) -> Result<String> {
+    Ok(format!("{:#X}", base58_to_move_address(base58)?))
+}
+
+/// Convert Move address to base58 address string
+/// Read more about the address structure: https://docs.substrate.io/reference/address-formats/
+///
+/// For now this function is supporting only one-byte address types, with one universal address type
+/// for all Move accounts and it's 42. We can't do otherwise as we've no information about the
+/// network we're currently compiling for.
+///
+/// Move developers should be aware, that information about any fails during the compilation of
+/// Move code will point to the addresses converted to the 42 address type.
+/// ```
+/// use move_core_types::account_address::AccountAddress;
+/// use move_vm_support::base58_address::move_address_to_base58_string;
+/// let move_address = "0x02A212DE6A9DFA3A69E22387ACFBAFBB1A9E591BD9D636E7895DCFC8DE05F331";
+/// let substrate_address = move_address_to_base58_string(&AccountAddress::from_hex_literal(move_address).unwrap());
+/// assert_eq!(
+///   "BH9wjrYwKi76DbU6LC1X4VEeevUTkaKgxGnaFNFaqZA",
+///  substrate_address
+/// );
+/// ```
+pub fn move_address_to_base58_string(addr: &AccountAddress) -> String {
+    let mut base58_address = [0; BASE58_LEN];
+    base58_address[0..PUB_KEY_LEN].copy_from_slice(&addr.to_vec());
+    bs58::encode(base58_address).into_string()
+}
+
+/// Converts a base58 address string to a Substrate SS58 address string.
+pub fn base58_string_to_ss588_string(addr: &str) -> Result<String> {
```

Comment:
not sure if we need these two functions here but who cares now since they are here 😃

Reply:
Same thoughts here 😆 👍